### PR TITLE
Fix printing of comma operators in object literals.

### LINKED ISSUE #3358: Comma operator is not correctly wrapped in () in object literals (computed keys and values)
Comma operators in the computed keys or in the values of an object literal are not correctly printed back to JavaScript source. They miss appropriate surrounding parentheses, causing `ParseError`s. This happens even in whitespace-only mode.

Reproduction:
```js
// ==ClosureCompiler==
// @compilation_level WHITESPACE_ONLY
// @output_file_name default.js
// @language_out ECMASCRIPT_2015
// @formatting pretty_print
// ==/ClosureCompiler==

function foo() {
  let array = [];
  const x = {
    [(array.push(1), "foo")]: (array.push(2), 3),
    [(array.push(3), "bar")]: (array.push(4), "foobar")
  };
  console.log(array);
  console.log(x);
}

foo(); foo();
```
is compiled with the above options as:
```js
'use strict';
function foo() {
  let array = [];
  const x = {[array.push(1), "foo"]:array.push(2), 3, [array.push(3), "bar"]:array.push(4), "foobar"};
  console.log(array);
  console.log(x);
}
foo();
foo();
```
[Appspot link](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520WHITESPACE_ONLY%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540language_out%2520ECMASCRIPT_2015%250A%252F%252F%2520%2540formatting%2520pretty_print%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Afunction%2520foo()%2520%257B%250A%2520%2520let%2520array%2520%253D%2520%255B%255D%253B%250A%2520%2520const%2520x%2520%253D%2520%257B%250A%2520%2520%2520%2520%255B(array.push(1)%252C%2520%2522foo%2522)%255D%253A%2520(array.push(2)%252C%25203)%252C%250A%2520%2520%2520%2520%255B(array.push(3)%252C%2520%2522bar%2522)%255D%253A%2520(array.push(4)%252C%2520%2522foobar%2522)%250A%2520%2520%257D%253B%250A%2520%2520console.log(array)%253B%250A%2520%2520console.log(x)%253B%250A%257D%250A%250Afoo()%253B%2520foo()%253B%250A)

Note the following problems:

* In a computed key: `[array.push(1), "foo"]` should be `[(array.push(1), "foo")]`
* In a value: `array.push(2), 3` should be `(array.push(2), 3)`

The problem can be reproduced
* with or without pretty-printing,
* in WHITESPACE_ONLY, SIMPLE_OPTIMIZATIONS and ADVANCED_OPTIMIZATIONS,
* tested with version v20190415

### PR DESCRIPTION
**Pull request by sjrd**
Both in a computed property key, and in value position.

Closes #3358.

### CODE REVIEW

**Comment by brad4d** [comment]
Created internal issue http://b/131904640

File: src/com/google/javascript/jscomp/CodeGenerator.java [hunk 1]
```diff
@@ -1061,7 +1061,7 @@ protected void add(Node n, Context context) {
           }
         }
         add("[");
-        add(first);
+        addExpr(first, NodeUtil.precedence(Token.COMMA) + 1, Context.OTHER);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/CodeGenerator.java]: Please revert this change, because parentheses aren't needed within the `[]` of the computed property.

File: src/com/google/javascript/jscomp/CodeGenerator.java [hunk 2]
```diff
@@ -1085,7 +1085,7 @@ protected void add(Node n, Context context) {
             checkState(
                 !isInClass, "initializers should only exist in object literals, not classes");
             cc.addOp(":", false);
-            add(initializer);
+            addExpr(initializer, NodeUtil.precedence(Token.COMMA) + 1, Context.OTHER);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/CodeGenerator.java]: This should be consistent with the behavior for a normal, STRING_KEY, type of property.
See method `addStringKey()`
tl;dr is to make this
`addExpr(initializer, 1, Context.OTHER)`

File: src/com/google/javascript/jscomp/CodeGenerator.java [hunk 3]
```diff
@@ -1085,7 +1085,7 @@ protected void add(Node n, Context context) {
             checkState(
                 !isInClass, "initializers should only exist in object literals, not classes");
             cc.addOp(":", false);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/CodeGenerator.java]: hmmm...
Given my statement on the next line to make this consistent with `addStringKey()`, I think this line should probably be
`cc.add(":")`

It's not technically correct to consider this colon an operator. I'm not sure it does any harm as it is currently written, but it does invoke a bunch of unnecessary logic.

File: test/com/google/javascript/jscomp/CodePrinterTest.java [hunk 4]
```diff
@@ -139,6 +139,15 @@ public void testObjectLiteralWithSpread() {
     assertPrintSame("({...{...{}}})");
   }
 
+  @Test
+  public void testObjectLiteralWithComma() {
```
Comment (brad4d) [inline test/com/google/javascript/jscomp/CodePrinterTest.java]: Well done remembering to add test cases.
Thanks.

Comment (sjrd) [reply test/com/google/javascript/jscomp/CodePrinterTest.java]: Ah ah thank you. That's actually the bare minimum for me. If there are other kinds of tests that would be relevant, I'm happy to add more if you can point me to the right place.

File: src/com/google/javascript/jscomp/CodeGenerator.java [hunk 5]
```diff
@@ -1061,7 +1061,7 @@ protected void add(Node n, Context context) {
           }
         }
         add("[");
-        add(first);
+        addExpr(first, 1, Context.OTHER);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/CodeGenerator.java]: Could you add a comment both here?

```java
// Must use addExpr() with a priority of 1, because comma expressions aren't allowed.
// LINK TO SPEC HERE
```

File: src/com/google/javascript/jscomp/CodeGenerator.java [hunk 6]
```diff
@@ -1084,8 +1084,8 @@ protected void add(Node n, Context context) {
             // Object literal value.
             checkState(
                 !isInClass, "initializers should only exist in object literals, not classes");
-            cc.addOp(":", false);
-            add(initializer);
+            cc.add(":");
+            addExpr(initializer, 1, Context.OTHER);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/CodeGenerator.java]: Could you add a comment here?
```java
// Must use addExpr() with a priority of 1, because a comma expression here would cause a
// syntax error within the object literal.
```

**Comment by brad4d** [comment]
Thanks for supplying the clarification and spec link on the issue.

If you'll just add the couple of comments I've requested, I should be able to import this PR.

Thanks.

**Comment by sjrd** [comment]
> If you'll just add the couple of comments I've requested

Done.

Thank you for the review. I'm afraid it probably ended up taking more or your time for you to review than implement yourself, though. Sorry about that :s

**Comment by brad4d** [comment]
imported and sent for internal review and testing

sorry for the delay

**Comment by brad4d** [comment]
submitted internally, should be in today's push to GitHub

**Comment by sjrd** [comment]
Thank you! Awesome contributing experience, btw.
